Fix osx and win32 builds

Tobias-Fischer opened the pull request:
On osx, one should not link against the uuid library - it's in the system's path already.

On win32, there are several fixes needed: `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` so a DLL is created, `add_definitions(-DNOMINMAX)` to avoid namespace clashes for `min` and `max`, and the uuid library there is simply called `Rpcrt4.lib`.

Tobias-Fischer committed:
Fix osx and win32 builds

traversaro commented:
Hi @mjcarroll, let us know if there is any we (as [RoboStack](https://robostack.github.io/), from which this patch originates: [osx](https://github.com/RoboStack/ros-galactic/blob/d4b28986669ca46ca4a73c7475c94f26060c7033/patch/ros-galactic-bondcpp.osx.patch) and [win](https://github.com/RoboStack/ros-galactic/blob/d4b28986669ca46ca4a73c7475c94f26060c7033/patch/ros-galactic-bondcpp.win.patch)) can do to help getting this PR merged, thanks!

Downstream issue: https://github.com/RoboStack/ros-galactic/issues/31 .

clalancette commented on bondcpp/CMakeLists.txt:
```diff
@@ -5,9 +5,12 @@ if(NOT CMAKE_CXX_STANDARD)
   set(CMAKE_CXX_STANDARD 14)
 endif()
 
+set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
While this probably does make it work on Windows, this is not the style that we have used in all of the rest of our packages.  In short, we explicitly add a `visibility_control.h` file to each of our projects, and then add the `RCLCPP_PUBLIC` (or whatever) on the symbols that need to be visible.  You can see an example of this at https://github.com/ros2/rcpputils/blob/master/include/rcpputils/visibility_control.hpp.  I'll suggest that we do the same here.

traversaro replied:
Ack, @Tobias-Fischer let me know if you prefer to do that or if you prefer that I do it, thanks!

Tobias-Fischer replied:
It would be great if you could do it @traversaro :)

traversaro replied:
Done in https://github.com/ros/bond_core/pull/83/commits/ea45bfbb4f0d0fe4bfa206793c9112664e531f98 . I also fixed `test_bond` compilation to make sure the symbols were correctly exported.

traversaro replied:
I realized I never mentioned @clalancette once I did this, so probably he missed this.

clalancette commented on bondcpp/CMakeLists.txt:
```diff
@@ -16,10 +19,14 @@ find_package(rclcpp_lifecycle REQUIRED)
+  # we have to find the absolute path to uuid as target_link_directories is not available before cmake 3.13
+  find_library(uuid_ABS_PATH ${UUID_LIBRARIES} PATHS ${UUID_LIBRARY_DIRS})
+elseif(WIN32)
+  set(uuid_ABS_PATH Rpcrt4.lib)
```
Does this really make it work on Windows?  Where does Rpcrt4.lib come from?

traversaro replied:
It is a system library: https://docs.microsoft.com/en-us/windows/win32/rpc/developing-32-bit-windows-applications .

clalancette commented on bondcpp/CMakeLists.txt:
```diff
@@ -16,10 +19,14 @@ find_package(rclcpp_lifecycle REQUIRED)
-pkg_check_modules(UUID REQUIRED uuid)
-# we have to find the absolute path to uuid as target_link_directories is not available before cmake 3.13
-find_library(uuid_ABS_PATH ${UUID_LIBRARIES} PATHS ${UUID_LIBRARY_DIRS})
+if(UNIX AND NOT APPLE)
```
What happens on macOS?  If we are never finding the uuid library, how does it build/link against it?

traversaro replied:
On macOS libuuid header and symbols are provided by the OS SDK/libc, so there is no need to provide additonal header directories or library to link, see https://github.com/ignitionrobotics/ign-cmake/pull/128 and https://github.com/ignitionrobotics/ign-cmake/issues/127 for a similar change on Ignition libraries or https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/uuid_generate.3.html for the macOS docs.

traversaro committed:
Explicitly mark symbol visibility and fix test_bond compilation on Windows

traversaro committed:
Fix definition of BONDCPP_BUILDING_LIBRARY

mjcarroll commented:
Howdy @traversaro , will take a look at this once I can get #87 landed to get a majority of the "cleanups" in.

Commit:
Merge branch 'ros2' into patch-8

traversaro commented:
> Howdy @traversaro , will take a look at this once I can get #87 landed to get a majority of the "cleanups" in.

Great, thanks!

mjcarroll commented on bondcpp/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ endif()
 if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   add_compile_options(-Wall -Wextra -Wpedantic)
 endif()
+add_definitions(-DNOMINMAX)
```
Do these definitions need to be added per-target rather than project wide?  That way they can be propagated to downstream.

Tobias-Fischer replied:
I've got no clue ;). @traversaro - do you have any input?

traversaro replied:
So, `-DNOMINMAX` in this case is used is required to make sure that `std::min` contained in the `bond.cpp` (https://github.com/ros/bond_core/blob/5a79fecb3e8b6f2c210d83669b5cb40d47d62c47/bondcpp/src/bond.cpp#L214) works fine, so fortunatly it is nothing in the public headers that needs to be exported. I guess this was added here to be coherent with the earlier use of of `add_compile_options` instead of `target_compile_options` , however I changed it to be per-target in https://github.com/ros/bond_core/pull/83/commits/141dbe1a9cade1112dd62f1a1153cd97e02a04ce .

mjcarroll replied:
Great thanks!

mjcarroll reviewed (commented):
I think this is mostly good, question on definitions, and I may have introduced a linting error on the merge.

mjcarroll commented:
@Tobias-Fischer friendly ping

traversaro committed:
Switch to use target_compile_definitions to define NOMINMAX

clalancette reviewed (approved):
Looks reasonable to me with green CI.

iggyrrieta commented:
Hi! Thanks for publishing this! 

I just cloned and failed building with following specs:

- ROS 2 Humble 
- Windows SDK version 10.0.22621.0 to target Windows 10.0.19045
- C compiler identification is MSVC 19.29.30152.0
- CXX compiler identification is MSVC 19.29.30152.0


Command I tried: 

```bash
colcon build --symlink-install --merge-install --packages-select bond smclib bondcpp --event-handlers console_direct+
```

Output:

- [x] bond
- [x] smclib
- [ ] bondcpp

> Summary: 2 packages finished [53.6s]
  1 package failed: bondcpp

clalancette commented:
After trying this locally, we definitely need a fix in `rclcpp_lifecycle` for this to work; see https://github.com/ros2/rclcpp/pull/2331 .  After that fix, there are some additional warnings in here:

```
C:/bond_ws/install/include/bond\bond/msg/detail/constants__struct.hpp(66,1): warning C4305: 'initializing': truncation from 'double' to 'const float' (compiling source file C:\bond_ws\src\bond_core\bondcpp\src\BondSM_sm.cpp) [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:/bond_ws/install/include/bond\bond/msg/detail/constants__builder.hpp(34): message : see reference to class template instantiation 'bond::msg::Constants_<std::allocator<void>>' being compiled (compiling source file C:\bond_ws\src\bond_core\bondcpp\src\BondSM_sm.cpp) [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:/bond_ws/install/include/bond\bond/msg/detail/constants__struct.hpp(66,1): warning C4305: 'initializing': truncation from 'double' to 'const float' (compiling source file C:\bond_ws\src\bond_core\bondcpp\src\bond.cpp) [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:/bond_ws/install/include/bond\bond/msg/detail/constants__builder.hpp(34): message : see reference to class template instantiation 'bond::msg::Constants_<std::allocator<void>>' being compiled (compiling source file C:\bond_ws\src\bond_core\bondcpp\src\bond.cpp) [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:\bond_ws\install\include\smclib/statemap.hpp(89,17): warning C4996: 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. (compiling source file C:\bond_ws\src\bond_core\bondcpp\src\BondSM_sm.cpp) [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:\bond_ws\install\include\smclib/statemap.hpp(89,17): warning C4996: 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. (compiling source file C:\bond_ws\src\bond_core\bondcpp\src\bond.cpp) [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:\bond_ws\src\bond_core\bondcpp\src\bond.cpp(561,53): warning C4244: '=': conversion from 'double' to 'bond::msg::Status_<std::allocator<void>>::_heartbeat_timeout_type', possible loss of data [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
C:\bond_ws\src\bond_core\bondcpp\src\bond.cpp(562,51): warning C4244: '=': conversion from 'double' to 'bond::msg::Status_<std::allocator<void>>::_heartbeat_period_type', possible loss of data [C:\bond_ws\build\bondcpp\bondcpp.vcxproj]
```

Once those are all fixed, we can consider merging this.

clalancette committed:
Merge branch 'ros2' into patch-8

clalancette committed:
Cast msg fields to float.

This is to avoid a warning on Windows.  It is safe to do
since the heartbeat timeout or period aren't expected to
be outside the range a float can represent.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
OK, we *also* need https://github.com/ros2/rosidl/pull/772 .  And I also just pushed some additional fixes.

With all of that in place, the last bit here is to fix the warnings about using `strncpy`.  Once that is fixed, I think this will be ready.

clalancette committed:
Revamp copyString to use memcpy.

Also make it allocate less memory in general.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>
